### PR TITLE
[WIP] Iss 2 flow2d write

### PR DESCRIPTION
**Pull request by MarcusUrban**
Please check the general approach.

### CODE REVIEW

File: mdal/frmts/mdal_flo2d.cpp [hunk 1]
```diff
@@ -675,3 +675,171 @@ std::unique_ptr< MDAL::Mesh > MDAL::DriverFlo2D::load( const std::string &result
...
+    }
+
+    for( size_t i = 0; i < timesCount; i++)
+    {
+        maximums.push_back(static_cast<float>(max));
+        minimums.push_back(static_cast<float>(min));
+        times.push_back(dsGroup->datasets[i]->time());
+    }
+
+    HdfGroup groupFlowDepth( groupTNOR.id(), "/TIMDEP NETCDF OUTPUT RESULTS/" + dsGroupName, true );
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_flo2d.cpp]: create 2 or 4 subfunctions: `store[Scalar/Vector][Face/Vertex]DatasetGroup` . one function should do 1 action

File: mdal/frmts/mdal_hdf5.cpp [hunk 2]
```diff
@@ -7,18 +7,31 @@
 #include <cstring>
 
 
-HdfFile::HdfFile( const std::string &path )
-  : d( std::make_shared< Handle >( H5Fopen( path.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT ) ) )
+HdfFile::HdfFile( const std::string &path, bool writeNew )
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_hdf5.cpp]: I do not like the new argument writeNew in the classes. I would prefer to either create the createFile/Attribute/... function outside the classes and then use the classes the old way?

In general in C++, you do not want constructors that could potentially fail/raise error.

File: mdal/frmts/mdal_hdf5.cpp [hunk 3]
```diff
@@ -194,22 +265,41 @@ std::string HdfDataset::readString() const
...
 
-HdfDataspace::HdfDataspace( const std::vector<hsize_t> &dims )
-  : d( std::make_shared< Handle >( H5Screate_simple(
-                                     static_cast<int>( dims.size() ),
-                                     dims.data(),
-                                     dims.data()
-                                   )
-                                 )
-     )
+void HdfDataset::writeString( hid_t fileId, hid_t dataspaceId, std::string value )
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_hdf5.cpp]: const std::string&

File: mdal/frmts/mdal_hdf5.hpp [hunk 4]
```diff
@@ -216,12 +222,25 @@ class HdfDataset
...
     float readFloat() const;
 
+    //! Reads string value
     std::string readString() const;
 
+    //! Writes string dataset and single data
+    void writeString( hid_t fileId, hid_t dataspaceId, std::string value );
+
+    //! Writes array of float data
+    void writeFloatArray( hid_t dataspaceId, float *value );
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_hdf5.hpp]: pass c++ objects, e.g const `std::vector<float>&`

File: mdal/frmts/mdal_flo2d.cpp [hunk 5]
```diff
@@ -675,3 +675,171 @@ std::unique_ptr< MDAL::Mesh > MDAL::DriverFlo2D::load( const std::string &result
...
+void MDAL::DriverFlo2D::saveNewHDF5File( DatasetGroup *group )
+{
+  // Create file
+  HdfFile file( group->uri(), true );
+
+
+
+  // Create dataspace for dataset File Version
+  std::vector<hsize_t> dimsSingle = {1};
+  float versionValue[] = {99.99f};
```
[the remark this answers is not in the document]
Comment (MarcusUrban) [reply mdal/frmts/mdal_flo2d.cpp]: sure, this is the value from the original file, we would need to read this value in a reader so we can increment it.

File: mdal/frmts/mdal_flo2d.cpp [hunk 6]
```diff
@@ -675,3 +675,171 @@ std::unique_ptr< MDAL::Mesh > MDAL::DriverFlo2D::load( const std::string &result
...
+
+  // Create attribute
+  HdfAttribute attTNORGrouptype( groupTNOR.id(), "Grouptype", true );
+  // Write string value to attribute
+  attTNORGrouptype.writeString( dscTNOR.id(), "Generic" );
+
+  for(auto dsGroup : group->mesh()->datasetGroups)
+  {
+    std::string dsGroupName = dsGroup->name();
+    if (dsGroupName == "Bed Elevation" || dsGroupName == "scalarGrp")
```
[the remark this answers is not in the document]
Comment (MarcusUrban) [reply mdal/frmts/mdal_flo2d.cpp]: No this is the content of the original file:
Screenshot from 2019-11-10 21-09-48
Then in the model there are 6 datasets loaded
0 - Bed Elevation
1 - FLOW DEPTH
2 - Floodplain...
3 - Velocity
4 - Velocity MAG
5 - scalarGrp which is empty

File: mdal/frmts/mdal_flo2d.cpp [hunk 7]
```diff
@@ -675,3 +675,171 @@ std::unique_ptr< MDAL::Mesh > MDAL::DriverFlo2D::load( const std::string &result
...
+
+    std::vector<float> maximums;
+    std::vector<float> minimums;
+    std::vector<double> times;
+
+    const size_t valuesSize = timesCount * dsGroup->mesh()->facesCount();
+    std::vector<float> values ( valuesSize );
+
+    size_t valuesIndex = 0;
+    if (dsGroupName != "Velocity")
```
[the remark this answers is not in the document]
Comment (MarcusUrban) [reply mdal/frmts/mdal_flo2d.cpp]: "Velocity" is the only vector dataset, I could use IsScalar, but then there is a DatasetUnits attribute which has different value for Velocity as well, so for sake of consistency I check for "Velocity". I dont see anything inside the file determining Face/Vertex data
Screenshot from 2019-11-10 21-27-56

File: mdal/frmts/mdal_flo2d.cpp (hunk 1 above)
Comment (MarcusUrban) [reply mdal/frmts/mdal_flo2d.cpp]: Agree, will do the in final version. Keeping it this way for now for simpler development and moving things around.

File: mdal/frmts/mdal_hdf5.cpp (hunk 2 above)
Comment (MarcusUrban) [reply mdal/frmts/mdal_hdf5.cpp]: I agree that this is far from ideal, I didnt want to break existing code and this kept other classes using Hdf classes working without a change. How to refactor this is probably for deeper discussion.

File: mdal/frmts/mdal_flo2d.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (PeterPetrik) [reply mdal/frmts/mdal_flo2d.cpp]: you cannot hardcode a string name of group in the writer/reader...

File: mdal/frmts/mdal_flo2d.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (PeterPetrik) [reply mdal/frmts/mdal_flo2d.cpp]: it is not part of the dataset group class. have you seen a file where the version is different than 99.99?

File: mdal/frmts/mdal_flo2d.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (PeterPetrik) [reply mdal/frmts/mdal_flo2d.cpp]: all dataset groups in FLO2D are on faces

[the remark this answers is not in the document]
Comment (PeterPetrik) [reply mdal/frmts/mdal_flo2d.cpp]: you cannot hardcode names in writer

**Pull request closed without merging**